ext/openssl/openssl.c

```diff
@@ -120,7 +120,13 @@ enum php_openssl_key_type {
 	OPENSSL_KEYTYPE_DH,
 	OPENSSL_KEYTYPE_DEFAULT = OPENSSL_KEYTYPE_RSA,
 #ifdef HAVE_EVP_PKEY_EC
-	OPENSSL_KEYTYPE_EC = OPENSSL_KEYTYPE_DH +1
+	OPENSSL_KEYTYPE_EC = OPENSSL_KEYTYPE_DH +1,
+#endif
+#if PHP_OPENSSL_API_VERSION >= 0x30000
+	OPENSSL_KEYTYPE_X25519 = OPENSSL_KEYTYPE_DH +2,
+	OPENSSL_KEYTYPE_ED25519 = OPENSSL_KEYTYPE_DH +3,
+	OPENSSL_KEYTYPE_X448 = OPENSSL_KEYTYPE_DH +4,
+	OPENSSL_KEYTYPE_ED448 = OPENSSL_KEYTYPE_DH +5,
```

Comment:
Why do you need new key types? Couldn't this just be part of EC supporting pub and priv key?

Reply:
So you want to divert from the openssl API and cram in  these curves in the current ECDSA parameters? EdDSA (with Ed25519/448) is not ECDSA. X25519/X448 is DH/key exchange only.

See https://mailarchive.ietf.org/arch/msg/cfrg/-9LEdnzVrE5RORux3Oo_oDDRksU/

Or other languages: https://cryptography.io/en/latest/hazmat/primitives/asymmetric/

Reply:
Some history can also be found here: https://crypto.stackexchange.com/questions/27866/why-curve25519-for-encryption-but-ed25519-for-signatures?rq=1

Reply:
That's not actually the best idea. Let's leave it separate...